compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -5260,6 +5260,144 @@ static bool skipRemLowering(int64_t divisor, TR::Simplifier *s)
+*    iconst 16
+*  iconst 16
+*
+*  into
+*
+* ificmpeq --> block_5
+*  iand
+*    iload  <...>
+*    iconst 16
+*  iconst 16
+*
+* Since the constants are equal, the case of greater-than is impossible,
+* and we can reduce ifcmpge to ifcmpeq. In other words, the result of iand
+* cannot be greater than the constant, and we can simply perform an equality check,
+* which provides further optimization opportunities.
+*
+* NOTE: the following special case.
+* We need to check that either the comparison is unsigned, or the mask has the
+* high/sign bit clear. Otherwise it's possible for the result of iand to compare greater.
+* For example, consider the following tree:
+*
+* ificmpge
+*   iand
+*     foo
+*     iconst -2
+*   iconst -2
+*
+* If foo is non-negative, then the result of iand will also be non-negative, and therefore > -2.
+*/
+class IfcmpgeToIfcmpeqReducer
+   {
+   public:
+   IfcmpgeToIfcmpeqReducer(TR::Node *node, TR::Simplifier *s) : _node(node), _s(s) {}
+
+   /**
+    * @brief Checks whether an ifcmpge node can be reduced to ifcmpeq.
+    *
+    * @return True if the reduction is possible.
+    */
+   bool reducible()
+      {
+      switch (_node->getOpCodeValue())
+         {
+            case TR::ifbcmpge:
+               return reducibleSpecific<int8_t>();
+            case TR::ifbucmpge:
+               return reducibleSpecific<uint8_t>();
+
+            case TR::ifscmpge:
+               return reducibleSpecific<int16_t>();
+            case TR::ifsucmpge:
+               return reducibleSpecific<uint16_t>();
+
+            case TR::ificmpge:
+               return reducibleSpecific<int32_t>();
+            case TR::ifiucmpge:
+               return reducibleSpecific<uint32_t>();
+
+            case TR::iflcmpge:
+               return reducibleSpecific<int64_t>();
+            case TR::iflucmpge:
+               return reducibleSpecific<uint64_t>();
+         }
+         return false;
+      }
+
+   /*
+    * @brief Perform the reduction.
+    *
+    * @return The transformed node.
+    */
+   TR::Node *reduce()
+      {
+      if (performTransformation(_s->comp(), "%sReduce an ifcmpge node [%p] to ifcmpeq\n", _s->optDetailString(), _node))
+         {
+         TR::ILOpCodes newOp = getIfcmpeqOpCode(_node);
+         TR::Node::recreate(_node, newOp);
+         }
+      return _node;
+      }
+
+   private:
+   /**
+    * @brief The actual implementation of `reducible` for an exact type.
+    */
+   template<typename T>
+   bool reducibleSpecific() const
+      {
+      TR::Node * firstChild = _node->getFirstChild(), * secondChild = _node->getSecondChild();
+
+      if (firstChild->getOpCode().isAnd() &&
+         firstChild->getSecondChild()->getOpCode().isIntegralConst() &&
+         secondChild->getOpCode().isIntegralConst())
+         {
+         T constAnd = firstChild->getSecondChild()->getConst<T>();
+         T constIfcmp = secondChild->getConst<T>();
+
+         bool isCmpUnsigned = _node->getOpCode().isUnsigned();
```

Comment:
I think the query you want here is `isUnsignedCompare`

Reply:
Fixed in 8f65726